### ISSUE TITLE
Idea - "Change State" Effect

### ISSUE THREAD

**Issue by 3LOK** [yelouafi/redux-saga#547]
Hi @yelouafi,

I'd like to preset an idea - if it'll sound good to you then I'll write it and add it as a pull request. The idea is to have a 'change state' effect, syntax as follows:

```javascript
    yield(changeState((state) => {
        const newState = {...state, x:1, y:2};
        return newState;
    });
```

This effect allows saga to directly change the state (and not go through reducers).

### Why is this good?
In our usage of sagas we do all logic (sync and async) in the sagas. At the end of each saga, the saga "puts" an action handled by the reducers that only updates the state to the value given. It makes the code, flows, and sagas much cleared - All business logics are centered in the sagas and sagas only. It makes it very clear to see flows, and to even test them (see my previous issue).

Basically, using this design, reducers have become unneeded 'dead weight'. Having this effect will remove the need for reducers (which, again, in our case do almost nothing).

### How would we use it?
Our system is very complex - it's an online ordering system, with its management dashboard used by hundreds of thousands of restaurants worldwide - including React Native apps that use the same sagas as the web. This design (all business logic in Sagas, no reducers whatsoever) made the system much easier to understand, develop, and test (we use TDD, or try to at least). 

What do you think?

**Comment by slorber** [comment]
What you want to achieve can easily be done by using a top-level reducer that listens for a `CHANGE_STATE` event, and replace the old state entirely by the new state. 

I'm not an expert in generator syntax but I'm sure it can be done with some code like:


```javascript
function* saga() {
  yield* changeState((state) => {...state, x:1, y:2});
}

function* changeState(update) {
   const state = yield select();
   yield put({type: CHANGE_STATE, data: update(state));
}

function liftReducer(reducer) {
  return (state, action) => {
    if ( action.type == "CHANGE_STATE" ) {
      return action.data;
    }
    else {
      return reducer(state,action)
    }
  }
}

const finalAppReducer = liftReducer(appReducer);
```

**Comment by slorber** [comment]
Also it looks to me a terrible idea.

The initial philosophy of a saga is to listen to events, manage its own internal state, and eventually emit new events.

The saga, in the first place, should generally avoid access to redux state. The `select` effect has been done because it is actually sometimes convenient and can help reduce boilerplate, but it is absolutly optional and for me it's not at all a best practice to rely on `select` heavily. The saga should be an autonomous component, that does not query other components of the system. Initially it comes from backend distributed systems. It is to consider that the saga is supposed to run on its own microservice, and should not communicate with other microservices with anything else that the event stream. It is generally advised, in distributed systems, that microservices do not access storage of one each others because it creates a lot of mess.

This is less dramatic for Redux, but still the same concerns apply. By relying on redux state inside sagas, and not only actions, you actually couple your saga to the state shape you choose for your React application. A properly designed saga should not assume anything about how your UI state is handled or shaped, but should just do its job and emit new events when needed. Actually, the saga should not even have to know that you are using React, or Redux. If you do things properly decoupled, you could switch to backbone and reuse the exact same unmodified sagas.

It is also important to emit new events, because these events can be useful for other sagas, or listened to by other pieces of the system (for example an event-tracking system like Mixpanel/GoogleAnalytics/Splunk/GetSentry...). If you modify state in place, it is much harder, later, to introduce these kind of systems. 

The saga somehow express "if This, then That". If you never emit an event for That, but only changes the state directly, nobody in the future will have the ability to plug behavior on the That event, that is never emitted.


You will also loose some ability for hot code reloading. By changing the state directly in the saga, it's not possible anymore to reinterpret your action log by running it again against an hot-reloaded updated reducer. Redux-saga in its current form will not help you much and it's quite hard to hot reload generators.

**Comment by 3LOK** [comment]
@slorber - Thanks for your very detailed answer. 

Most of apps contain different business logic flows. For example: 'When a user wants to checkout, show him a contact form, wait for him to click next, then show him a delivery form, ...', or for example: 'when a user wants to delete an item, show him a are-you-sure popup, wait for him to click yes or no, if he/she clicked no, close the popup. if they clicked yes, show a thinker, send the server request, and return back'. Sagas lets us write these flows very explicitly clear:

```javascript
    function* handleOnDelete(dep, action) {
        yield put(actions.showAreYouSurePopup());
        const result = yield take([types.DELETE_POPUP.YES, types.DELETE_POPUP.NO]);
        yield put(actions.hideAreYouSurePopup());
        if (result.type === types.DELETE_POPUP.YES) {
            yield put(actions.showThinking());
            yield call(dep.server.deleteItem, action.itemId);
            yield put(actions.stopThinking());
        }
    }
```

Up until here, I believe we both agree in the usage of sagas as a descriptive way to write your business logic. It's a very powerful tool, that can easily be read and tested. (correct me if I'm mistaken and you disagree). 

However, here's where I personally am willing to "pay" in all the advantages you've mentioned, in order to achieve something else: readability. Instead of throwing actions that change the state using reducers (even the generic reducer you suggested), I want to change the state in the saga. And furthermore, in some sagas i DO want to access the state (which tightly couples the saga to Redux state). 

Why? Because I feel that the advantage of having your entire business logic flow concentrated in one place makes development, debugging, and refactoring of business logic ten times easier. To be honest, I believe the 'payment' in complexity I'm paying navigating between different files and flows due to logic split between reducers and sagas accumulates to much more (in terms of developer hours) than the payment I would pay if I ever want to refactor my code from Redux to Backbone (I'll probably do that once every 2-3 years, compared to debugging I do all the time).

So to comment specifically on things you said:
1. Saga not accessing other storage - I agree, but I view it differently. For me there's the constant pairing of <Saga, Reducer>, and I feel that each pair should have a single 'database' (the specific path in the store). I don't believe pairs should access other pairs 'storage', but I think in the pair it should be perfectly valid and logical to have the saga access the store it manipulates.

2. Coupling sagas and redux states - again, I'm willing to pay in this coupling to have clearer and easier code to follow and debug. I don't really think we will move away from redux without also changing the sagas on one hand, and at the same time I'm willing to let "Yoav of 2018" pay the price for this decision so "Yoav of 2016" will have an easier time following the flow of the system.

3. Events for other sagas/plugins - I agree - I'll be honest that this point is making me reconsider my solution. Let me think about this :) I might design a special structure of events that you can explicitly define how they need to change the state.

4. Hot reloading - I think when you move to redux-saga you generally release any hot-reloading of your business logic, and focus your hot reloading only on your React layer. I don't think my structure makes hot reloading less feasible on the "Flux" level then it is right now.

Again, thanks for your comments.

**Comment by slorber** [comment]
Hi,

I understand your point. As far as I understand your delete button is a lot coupled to your delete confirmation popup, and emitting open/close events, and having to maintain a reducer for the popup feel like a lot of complexity.

In distributed systems you could say these 2 components are part of the same microservice (or bounded context) that handles the whole deletion of your item. Then if it's perfectly clear that the 2 components and the saga act inside the same bounded context, I think it's fine if all 3 knows about the state shape. I've done that in the past too. 

Your bounded context of the item deletion is composed of a delete button, a delete popup, and a saga that orchestrate the button with the popup. In my app, I call this a "widget" and give to this widget a subspace of my Redux store. Ideally, only this widget can `select` or `connect` state from its store subspace. The widget act somehow like a microservice which has its own private/internal database, and it's database is actually mounted inside the Redux store. This way we preserve the decoupled advantages.

--------------------------------------------------

However, for your specific usecase, it's very rare that we need to listen to confirmation popup open/close events from other parts of your system. Generally these operations are actually to consider as implementation details of your widget, and maybe it's not worth it to expose them as widget API.

What I mean is, in my opinion, if you are concerned with boilerplate, it's really much easier to handle the delete confirmation popup and the spinner with React local component state. Yes, you will loose the ability to hot reload the confirmation modal and the spinner, but it really isn't a big deal imho unless it's very important to you.

If you want the confirmation popup to be appended to the document body (for z-index reasons), I'd recommend using a [portal](https://github.com/tajo/react-portal) which is very useful to reduce this kind of boilerplate.

For example you could have:

```javascript
React.createClass({
  getInitialState() {
    return { confirmationPopup: false };
  },
  render() {
    return (
       <div className="delete-item-widget">
          <div onClick={() => this.setState({confirmationPopup: true})}>Delete item</div>
          {this.state.confirmationPopup && (
              <Portal>
                 <ConfirmationPopup 
                      onCancel={() => this.setState({confirmationPopup: false})}
                      onConfirm={() => this.props.onDelete())}
                 >
              </Portal>
          )}
       </div>
    );
  }
})
```

Even if the portal is deeply nested, it gets appended to document body.

For "positionned portals" you can use [react-tether](https://github.com/souporserious/react-tether) which is extremely convenient and permit to avoid the mess of relative positionning and z-index management.

For spinner, you can also use local component state, and make your actions return promises. It can also be helpful if you make the prop functions (onDelete) return the deletion promise, so that deeply nested components can eventually decide to show a spinner.

Doing all this will generally reduce a lot the boilerplate of your application, and in my experience the tradeoff for using these shortcuts is not a big deal

**Comment by yelouafi** [comment]
@3LOK I don't really have a strong opinion against bypassing reducers. Perhaps many Redux users will find this *shocking*, but I've seen this quite often in FRP examples (RxJS, Bacon) where we produce a stream of update functions which get merged and applied to some central state (Maybe because in those systems there is no explicit concept of action).

But I think this is beyond the scope of redux-saga as a middleware. Even If we want we can't actually because there is no way for a middleware to change state in the store; In fact, the only way to change state in Redux is through dispatching actions to reducers. But then you can implement this by yourself by dispatching a specific action 'CHANGE_STATE' which will be handled by a generic reducer

```js
function genericReducer(state, action) {
  return action.updateFn(state)
}

// make an Effect creator
const changeState = updateFn => put({ type: 'UPDATE_STATE', updateFn })

function* saga() {
  yield changeState(state => ...)
}
```

But I think the issues raised by @slorber remains pertinent.  Sagas are fundamentally Event driven; and the advantage of dispatching actions is that an action can be handled by other Sagas.

It may be not obvious because as @slorber mentioned, in your application you have so many *bounded contexts* (logic that is local to a specific component). 

The main benefit of Redux (and Flux solutions in general) is to externalize some global piece of logic: I mean some piece of state used by multiple UI components and also can be updated as a  side effect of many local updates ([The Facebook Chat example](https://www.youtube.com/watch?list=PLb0IAmt7-GS188xDYE-u1ShQmFFGbrk0v&v=nYkdrAPrdcw) is a good illustration). So if in your applications you dont have much of this global logic the Redux architecture (actions/reducers) may seem an unnecessary overhead to use for all local state updates.

**Comment by yelouafi** [comment]
>Events for other sagas/plugins - I agree - I'll be honest that this point is making me reconsider my solution. Let me think about this :) I might design a special structure of events that you can explicitly define how they need to change the state.

You may want to take at this article

[Using React (-Native) with Redux and Redux-Saga. A new proposal?](https://medium.com/@marcelschulze/using-react-native-with-redux-and-redux-saga-a-new-proposal-ba71f151546f#.mpuwlwfeo)

If you abstract away firing actions whose specific purpose is to update the state, and replace them by updating the state directly (by dispatching instead functions which will update the state) you may find the purpose similar to you proposal.
